### PR TITLE
React 16 upgrade

### PR DESCRIPTION
**Pull request by alexreardon**
There is still a lot to do on this one. But I am starting the PR early. Closes #406 #192 #103 #8 

- [x] Upgrade react dependencies
- [x] Upgrade enzyme and ensure all tests are passing
- [x] Fix linting
- [x] Convert placeholder to use the same element type as the element dragging
- [x] ~~Move to portals~~
- [x] Thrash out new placeholder api
- [x] Update docs and examples to use new `Draggable` api
- [x] Add portal example
- [x] Add a portal guide
- [x] Add tests for post portal movement auto focus (and focus publishing)
- [x] add tests for new behaviour (if any)
- [x] Ensure all tests are passing
- [x] Build sortable table example
- [x] Fix linting
- [x] Fix flow
- [x] Write table sorting guide
- [x] Update all stories and docs to use new api
- [x] Write release notes (including perf gain table 😲)

In scope?
- [x] Moving to React 16.3
- [x] Using the new React 16.3 lifecycle methods (or at least moving off the deprecated ones)
- [ ] Internally move to the new Context api

Out of scope
- Placeholder animations #77 (can do later)

Currently stating min version of React 16.3. ~~Should this be increased to 16.3? At this stage I am not interested in being slowed down by trying to be compatible with older React versions 🤔~~

### CODE REVIEW

**Comment by tim-soft** [comment]
I can't imagine many situations where someone may be "stuck" from upgrading from React 16.2 -> 16.3.
 
That upgrade path is a much smaller burden on the user than the bottleneck of being backward compatible. If you are looking for opinions I say go all the way ¯\\\_(ツ\)_/¯

**Comment by alexreardon** [comment]
Agreed @tim-soft

**Comment by alexreardon** [comment]
screen shot 2018-04-05 at 3 14 33 pm

I have moved everything over to React 16.3 including moving away from the deprecated lifecycle methods. Making progress!

File: docs/patterns/tables.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# Tables
+
+| Benefits of using `<table>`                      | Provider                |
+|--------------------------------------------------|-------------------------|
+| Clean way of displaying tabular data             | Browser                 |
+| Great browser support                            | Browser                 |
+| Can copy paste the table into other applications | Browser                 |
+| Can reorder items in the table!                  | `react-beautiful-dnd` 😎|
+
+`react-beautiful-dnd` supports requires no additional wrapping elements to create `Draggable` and `Droppable` components. Therefore it is possible to have a `<table>` that has valid HTML as well as supporting drag and drop.
```
Comment (petegleeson) [inline docs/patterns/tables.md]: "`react-beautiful-dnd` requires no additional"

File: docs/patterns/tables.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+1. Fixed layouts (faster and simplier)
+2. Dimension locking (slower but more robust)
+
+### Strategy 1: fixed layouts
+
+In order to use this strategy the widths of your columns need to be fixed - that is, they will not change depending on what content is placed in the cells. This can be achieve with either a `table-layout: fixed` or `table-layout: auto` as long as you manually set the width of the cells (eg `50%`).
+
+The only thing you need to do is set `display: table` on a `Draggable` row while it is dragging.
+
+[See example code here](TODO)
```
Comment (petegleeson) [inline docs/patterns/tables.md]: Do we have code samples for here, [here](https://github.com/atlassian/react-beautiful-dnd/pull/406/files#diff-8029e14a5c1d3dab700f1fc881dcedafR42) and [here](https://github.com/atlassian/react-beautiful-dnd/pull/406/files#diff-8029e14a5c1d3dab700f1fc881dcedafR66)?

File: stories/1-single-vertical-list-story.js [hunk 3]
```diff
@@ -8,7 +8,7 @@ import { grid } from './src/constants';
 
 const data = {
   small: quotes,
-  // small: getQuotes(3),
+  // small: quotes.slice(0, 2),
```
Comment (petegleeson) [inline stories/1-single-vertical-list-story.js]: can we remove this?

File: stories/src/primatives/quote-item.jsx [hunk 4]
```diff
@@ -107,9 +107,15 @@ export default class QuoteItem extends React.PureComponent<Props> {
...
     node.focus();
+    // console.log('quote item mounted', this.props.quote.id);
+  }
+
+  componentWillUnmount() {
+    // console.log('quote item unmounted', this.props.quote.id);
   }
 
   render() {
+    // console.log('quote item rendered', this.props.quote.id);
```
Comment (petegleeson) [inline stories/src/primatives/quote-item.jsx]: Can probably remove the changes to this file

File: stories/src/table/with-dimension-locking.jsx [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+  isDragOccurring: boolean,
+|}
+
+type TableCellSnapshot = {|
+  width: number,
+  height: number,
+|}
+class TableCell extends React.Component<TableCellProps> {
+  // eslint-disable-next-line react/sort-comp
+  ref: ?HTMLElement
```
Comment (petegleeson) [inline stories/src/table/with-dimension-locking.jsx]: Can use [ElementRef](https://flow.org/en/docs/react/types/#toc-react-elementref) to type these variables.

File: stories/src/table/with-portal.jsx [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+Object.assign(table.style, {
+  margin: '0', padding: '0', border: '0',
+});
+const tbody: HTMLElement = document.createElement('tbody');
+table.appendChild(tbody);
+
+if (!document.body) {
+  throw new Error('document.body required for example');
+}
+document.body.appendChild(table);
```
Comment (petegleeson) [inline stories/src/table/with-portal.jsx]: Will this add many tables because of HMR while in dev?

**Comment by alexreardon** [comment]
Yeah it will. But I am not worried about it. You think it is worth cleaning up the portals?

File: docs/patterns/tables.md (hunk 2 above)
Comment (alexreardon) [reply docs/patterns/tables.md]: We do now! 👍

Comment (alexreardon) [reply docs/patterns/tables.md]: I'll update the links

File: stories/src/table/with-dimension-locking.jsx (hunk 5 above)
Comment (alexreardon) [reply stories/src/table/with-dimension-locking.jsx]: These are actualy HTMLElement and not component instances

File: stories/src/table/with-portal.jsx (hunk 6 above)
Comment (alexreardon) [reply stories/src/table/with-portal.jsx]: For now I think it is fine

**Pull request merged**